Components: Refactor SlotFill

diegohaz opened the pull request:
## Description

This PR is part of #18619.

### The problem

There are several problems with the current `slot-fill` implementation. For example, the context value never changes because it's a class property whose value is never updated during the `SlotFillProvider` lifecycle:

https://github.com/WordPress/gutenberg/blob/56cd1df587d48fcb7d710aaae12098832a40a152/packages/components/src/slot-fill/context.js#L38-L47

Because of that, both the `Slot` and the `Fill` components will never re-render when some action is done in the context (for example, when calling `registerSlot` or `registerFill`). Workarounds like `forceUpdate()` have been introduced to the code to overcome this:

https://github.com/WordPress/gutenberg/blob/56cd1df587d48fcb7d710aaae12098832a40a152/packages/components/src/slot-fill/context.js#L55-L65

Also, the `Fill` component won't re-render when new `fillProps` are passed to the `Slot` component (probably, a new `forceUpdate` call should be placed here to work around this issue):

https://github.com/WordPress/gutenberg/blob/56cd1df587d48fcb7d710aaae12098832a40a152/packages/components/src/slot-fill/fill.js#L56-L59

### This solution

This is a rewrite of the `slot-fill` module using React Hooks and exposing only the Portal API (activated by using the `bubblesVirtually` prop today). This greatly simplifies the code without workarounds.

The component should behave exactly the same as when using the `bubblesVirtually` prop today, except that it will properly re-render when `fillProps` are updated.

The idea is to gradually migrate the codebase to use this while this is experimental and then replace the current `slot-fill` with this one if the experiment succeeds. This would also validate whether the current usage of slots without `bubblesVirtually` could use it or not.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

This was tested using Storybook and unit tests.

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

diegohaz committed:
SlotFill initial implementation

diegohaz committed:
Add manifest-devhub.json

diegohaz committed:
Accept as prop on Slot

diegohaz committed:
Update stories

diegohaz committed:
Update README.md

diegohaz committed:
Update code

diegohaz committed:
Add slot-fill2 entries to components index file

diegohaz committed:
Add unit tests

diegohaz commented on packages/components/src/slot-fill2/context.js:
```diff
@@ -0,0 +1,94 @@
+export const SlotFillContext = createContext();
+
+function useSlotRegistry() {
+	const [ slots, setSlots ] = useState( {} );
```
On master, we have a `getSlots` function whose reference will never change. By putting the `slots` object into state directly, this will automatically trigger an update/re-render on consumers (`Slot` and `Fill` components) when new slots are added, removed or updated (`fillProps`).

gziolo replied:
This is neat. I like this proposal. It removes the need for the existence of the `occurrences` local variable.

diegohaz commented on packages/components/src/slot-fill2/fill.js:
```diff
@@ -0,0 +1,23 @@
+import { useSlot } from './context';
+
+export default function Fill( { name, children } ) {
+	const slot = useSlot( name );
```
This will trigger a re-render on this component when:
- a new slot has been added.
- a slot has been updated (`fillProps` or `ref`).
- a slot has been removed.

diegohaz commented on storybook/test/__snapshots__/index.js.snap:
```diff
@@ -3444,6 +3444,54 @@ exports[`Storyshots Components|ScrollLock Default 1`] = `
+  </p>,
+]
+`;
+
```
It looks like storyshots doesn't support React Portal? :(

gziolo replied:
I wouldn't be surprised if you would have to mock it somehow. See also:

https://github.com/WordPress/gutenberg/blob/0f2ec15ed40192a165f79790effa8edaf6504597/storybook/test/index.js#L9-L27

gziolo replied:
I tried to run old unit tests that exist for the slot component:

https://github.com/WordPress/gutenberg/blob/master/packages/components/src/slot-fill/test/slot.js

They all fail, so it might be an actual issue with the implementation. Well, many of those tests are for the case that doesn't bubble virtually, so it's hard to tell :)

diegohaz replied:
I moved `slot-fill2` into `slot-fill` and replaced the current `bubblesVirtually` implementation with it. I had to remove only one assertion that was testing the component instance, but the rest is working fine. I'll add more tests later.

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz committed:
Fix git conflicts

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz committed:
Lint code

gziolo commented on packages/components/src/slot-fill2/index.js:
```diff
@@ -0,0 +1,23 @@
+/**
```
What about the set of tests for the slot implementation? I don't see them ported.

I even copied them and tried to run locally but they all fail.

diegohaz replied:
I moved `slot-fill2` into `slot-fill` and replaced the current `bubblesVirtually` implementation with it. I had to remove only one assertion that was testing the component instance, but the rest is working fine. I'll add more tests later.

gziolo reviewed (commented):
It would be great to port more existing unit tests first (wherever possible).

The next step would be to replace the old implementation for `bubblesVirtually` with this proposal. If it works without issues we can land it. What do you think?

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz committed:
Update manifest.json

diegohaz committed:
Try: replace <Slot bubblesVirtually /> by Slot2

diegohaz committed:
Set a default value for SlotFillContext

diegohaz committed:
Update SlotFillContext slots initial value

diegohaz committed:
Refactor code

diegohaz committed:
Update docs/manifest.json

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -69,21 +72,15 @@ const BlockInspector = ( {
+					>
+						<InspectorAdvancedControls.Slot bubblesVirtually />
+					</PanelBody>
+				) }
```
Related discussion: https://github.com/WordPress/gutenberg/pull/16807#pullrequestreview-268314678

diegohaz commented on packages/components/src/index.js:
```diff
@@ -98,7 +98,7 @@ export {
 	Fill,
 	Provider as SlotFillProvider,
-	Consumer as __experimentalSlotFillConsumer,
+	useSlot as __experimentalUseSlot,
```
If I got it right, we only had to expose `Consumer` for the cases where we needed to check if slots or fills have been registered.

On `Popover`, we check the presence of the slot in order to render the contents within a `<Fill>`:
https://github.com/WordPress/gutenberg/blob/59b07cc57adff90630fc9d5cf2317269a0f4f158/packages/components/src/popover/index.js#L606-L622

On `BlockInspector`, we check the presence of fills so we can conditionally render the slot wrapper:
https://github.com/WordPress/gutenberg/blob/59b07cc57adff90630fc9d5cf2317269a0f4f158/packages/block-editor/src/components/block-inspector/index.js#L72-L86

youknowriad replied:
Yes, I like the hook better, nice work here.

diegohaz commented on packages/components/src/popover/index.js:
```diff
@@ -602,25 +603,15 @@ const Popover = ( {
-			} }
-		</Consumer>
-	);
+	return <span ref={ anchorRefFallback }>{ content }</span>;
```
[Looks like PopoverSlot always uses bubblesVirtually](https://github.com/WordPress/gutenberg/blob/0b11256824a1aae503ade5d7071c42d2b2c354ce/packages/components/src/popover/index.js#L619-L621), so I also changed this snipper to use `useSlot` instead of `Consumer`. With that, `Consumer` is only used on the `SlotFill` module internally.

diegohaz commented on packages/components/src/slot-fill/bubbles-virtually/fill.js:
```diff
@@ -0,0 +1,31 @@
+		return () => {
+			slot.unregisterFill( ref );
+		};
+	}, [ slot.registerFill, slot.unregisterFill ] );
```
This is just so we can keep track of fills and check their existence later (on `BlockInspector`, for example).

diegohaz commented on packages/components/src/slot-fill/index.js:
```diff
@@ -1,13 +1,28 @@
+			<BubblesVirtuallyFill { ...props } />
+		</>
+	);
+}
```
This probably needs some code comments. The reason I'm including both `Fill`s there is because we have to register them even before their respective slot is registered (so `BlockInspector` can check if there are fills before rendering the slot, for example).

They both registers themselves and return `null` if there's no slot. So, one of them will be always `null`.

diegohaz commented on packages/components/src/slot-fill/test/slot.js:
```diff
@@ -7,9 +7,7 @@ import ReactTestRenderer from 'react-test-renderer';
-import Slot from '../slot';
-import Fill from '../fill';
-import Provider from '../context';
+import { Slot, Fill, Provider } from '../';
```
Will add more tests

diegohaz commented on packages/components/src/slot-fill/test/slot.js:
```diff
@@ -261,10 +259,6 @@ describe( 'Slot', () => {
-
-				expect( testRenderer.getInstance().slots ).toHaveProperty(
-					'egg'
-				);
```
I can't really reproduce this with a function component. I'll add more tests.

diegohaz commented on packages/components/src/slot-fill2/README.md:
```diff
@@ -0,0 +1,66 @@
+
+The `SlotFillProvider` component does not accept any props.
+
+Both `Slot` and `Fill` accept a `name` string prop, where a `Slot` with a given `name` will render the `children` of any associated `Fill`s.
```
I removed this README.md file as the contents of `slot-fill2` were moved over to `slot-fill`. But, the reason to remove that was because `<Slot bubblesVirtually />` didn't accept children prop. Since this refactor only considers that use case, it didn't make sense keeping that in the readme.

But now it's possible to access fills through `useSlot`. See https://github.com/WordPress/gutenberg/pull/19242#discussion_r381127219

diegohaz committed:
Update story title separator

diegohaz commented:
Thank you so much for the review @gziolo! ❤️

So, I've moved the `slot-fill2` module into the current `slot-fill` one and replaced the current `bubblesVirtually` implementation with it. This also means that the current tests are also testing the refactored implementation. Although I think it needs more tests.

But CI is passing and the code is ready for another review.

youknowriad commented on packages/components/src/slot-fill/bubbles-virtually/fill.js:
```diff
@@ -0,0 +1,31 @@
+ */
+import useSlot from './use-slot';
+
+export default function Fill( { name, children } ) {
```
This is a simple and nice implementation but I believe we had issues with fills being reordered when we switch between blocks (toolbar and inspector) and we had to add a custom logic to keep the fills rendering in the exact same order they appear on the React tree (even with conditions).

I see that this new implementation passes the existing unit tests right which I believe account for this (and I'm surprised it does without any tweaks to the rendering order). We might just want to confirm in the UI too.

diegohaz replied:
Hmm, from what I can see, the current unit tests are not covering that case for `bubblesVirtually`. Actually, it seems that `bubblesVirtually` is not being covered at all? `react-test-renderer` doesn't seem to support portals. But I'll check if `react-dom/test-utils` can do the job.

Also, I can't find how it's been solved for the current implementation (using `bubblesVirtually`). Do you have an idea?

I'm pretty sure that my implementation doesn't cover that. `ReactDOM.createPortal`, by default, will just attach elements in the order they're rendered.

I've just made some experiments with the `key` prop here: https://codesandbox.io/s/keeping-react-portal-order-uk9xb

youknowriad replied:
It's related to the `occurrence` counter we have on fills and there's  a function to reorder the fills in the Provider I believe. That said, If we manage to run the unit tests on the new implementation, that would be enough to validate it I think.

youknowriad reviewed (commented):
Looks great, I have a small concern about the fills order (we just need to check) and I agree with the plan of deprecating the non bubblesVirtually slot/fill entirely. This was always the intent but we didn't complete the refactor.

Which slots still use the non-bubbles virtually version?

diegohaz committed:
Update snapshots

youknowriad commented on packages/components/src/slot-fill/index.js:
```diff
@@ -1,13 +1,28 @@
-import Fill from './fill';
-import Provider, { Consumer } from './context';
+import BaseSlot from './slot';
+import BaseFill from './fill';
```
Should we remove all the bubblesVirtually logic from the BaseSlot, Baseffill and Provider?

diegohaz committed:
Remove bubblesVirtually implementation from BaseFill/Slot

diegohaz committed:
Make sure fills are being created in the right order

diegohaz committed:
Merge branch 'master' into update/slot-fill

diegohaz committed:
Add comment on Fill dual rendering

diegohaz committed:
Add code comments on Fill

diegohaz committed:
Try with compareDocumentPosition

diegohaz committed:
Revert ordering feature

diegohaz commented:
So, it turns out that both the current code on master that uses `bubblesVirtually` and this PR don't cover the case where fills could be mounted asynchronously in different order.

I tried different approaches to solve that:

### Re-mount the portal when its position in the DOM had to be corrected

  This can be achieved by passing a different `key` to the portal:
  
  ```jsx
  createPortal( element, container, key );
  // or
  createPortal( <Fragment key={ key }>{ element }</Fragment>, container );
  ```

  To identify when the `key` should be changed, I tried different methods:

  - Render an empty `<span />` in place on the `Fill` component and use [`Node.compareDocumentPosition`](https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition) to determine where to add the fill in the array by comparing its position with the other fills' spans in the DOM. [This is what we use on Reakit](https://github.com/reakit/reakit/blob/88224725f4299f282cdaec357d2ed7b5e6c5ba70/packages/reakit/src/Rover/RoverState.ts#L140-L166) so we can add items to an array in the same order they're rendered in the DOM.

  - Only change the `key` when fills that were already registered tried to register themselves again, which would indicate a new render cycle. If the new fills array were different from the last cycle, that was because a fill has been added or removed.

  - A simpler solution: change `key` whenever the `fills` array changes. It resulted in several re-mounts.

  According to the [tests](https://github.com/WordPress/gutenberg/blob/aa6160521568848f4bec19510c569c8aa53427c5/packages/components/src/slot-fill/test/slot.js#L224-L275), all of those methods successfully kept fills in the expected order. But in all those cases the block toolbar behaved quite weirdly:

  ![Feb-22-2020 22-21-51](https://user-images.githubusercontent.com/3068563/75104296-58ddee80-55e6-11ea-99a6-24c48e924366.gif)

  This clearly had something to do with the portal re-mounting, but I couldn't figure out how to mitigate that. Maybe it's something specific to how block toolbars responds to mouse over and out? Tried more variations, but I was spending too much time on this `key` strategy. So I decided to research more and try other things.

### Instead of moving the portal itself, put it inside a `<span>` and work only on that

  ```jsx
  const [ span ] = useState( () => createElement( 'span' ) );
  // it'll be re-appended on every render so it'll always be in the correct position 
  useEffect( () => {
    slotNode.appendChild( span );
    return () => span.remove();
  } );
  // createPortal on `span` instead of `slotNode`
  return createPortal( children, span );
  ```

  This is based on a [comment by Dan Abramov](https://github.com/facebook/react/issues/12247#issuecomment-486378549). The issue they're trying to solve there is not the same, but this was the solution that, after a few tweaks, had the best result. As long as you keep the reference to the container element, you can move it freely and the portal element will not need to be recreated.

  Still, this presented a lot of problems: fill elements now would have a `<span>` wrapping them, which broke the styling. Adding `flex: inherit` and other css rules to the span fixed the issue. But, for some reason, now you could have two popovers open at the same time within the block toolbar. And some e2e tests were failing.

### Conclusion

I ended up reverting those attempts to fix the order issue. It looks like, at least for the modules that are currently using `bubblesVirtually`, they don't need this kind of precision?

It took so many workarounds for this to vaguely work that I think we should, in the future, re-consider the `bubblesVirtually` approach altogether. Maybe using React Portal wasn't a good idea? After all, it's not even supported by React Native.

This PR is already an improvement over the current implementation on master. It properly updates fills when `fillProps` change and gets rid of the `Consumer`. So I'm still in favor of landing it as is.

youknowriad commented:
So I do think we need this precision at some point but I agree that the fact that it's present in master makes it less important to fix in this particular PR.

youknowriad reviewed (approved):
This is looking great, I'd appreciate a follow-up to switch all slots to the new implementation and officially deprecate/remove the old implementation.

I'd also appreciate @aduth thoughts here (even after merge as he's AFK).

gziolo commented:
There is one blocker to resolve:

![Screen Shot 2020-02-24 at 13 23 10](https://user-images.githubusercontent.com/699132/75152225-d8fc7500-5708-11ea-9587-2f90853828b7.png)

When you enable the top toolbar option in the More Menu, most of the controls are missing.

gziolo commented on packages/e2e-tests/specs/editor/various/writing-flow.test.js:
```diff
@@ -512,9 +512,9 @@ describe( 'Writing Flow', () => {
-		await page.waitForSelector(
-			'.block-editor-block-list__insertion-point-inserter'
-		);
+		// await page.waitForSelector(
```
Why is it commented out? It doesn't seem like it should have any impact on toolbars.

diegohaz replied:
Ah, I forgot it there. That testing was failing while I was trying to fix the ordering issue. I was playing with the test code to see if there was some problem in the test implementation. Just reverted this.

gziolo commented on packages/block-editor/src/components/block-inspector/index.js:
```diff
@@ -30,6 +30,9 @@ const BlockInspector = ( {
 	selectedBlockName,
 	showNoBlockSelectedMessage = true,
 } ) => {
+	const slot = useSlot( InspectorAdvancedControls.slotName );
```
Yes, let's keep `__experimentalUseSlot` experimental until we come up with a better way to handle it. It seems fine for the time being.

diegohaz committed:
Uncomment test

diegohaz committed:
Fix top toolbar not updating properly

diegohaz commented:
Good catch @gziolo! Thanks for the review! ❤️

So, I found the problem and fixed it. This is something that should be covered by the current tests (`should subsume another slot by the same name`). But `react-test-renderer` doesn't support portals (https://github.com/facebook/react/issues/11565), so even on master that is not being covered.

I'll propose in a follow-up PR a refactored set of tests for this module using react-testing-library (using the dependency directly, without the `@wordpress/test-utils` package proposed in #18855).

diegohaz commented:
Follow-up PR with improved tests: #20428

gziolo reviewed (approved):
I can confirm it fixes the issue I raised yesterday. Awesome work on fixing it. I believe it closes more existing issues. I will try to identify them :)

Let's ship it 🚢

aduth commented:
This looks really cool! I'd like to circle back and dive into the specific changes, but at least on a cursory glance, the implementation looks solid.

Would you happen to have a sense whether this will help unblock #14845 / #17355 ? I noticed some earlier comments at https://github.com/WordPress/gutenberg/issues/17355#issuecomment-582270165 , but those were prior to some additional changes that came later in this pull request.

draganescu commented:
Advised by @aduth I tried reverting this PR and now this #18560 new editor crash is fixed. So this PR apparently did something that crashes the editor with the console message:

`TypeError: "slot is undefined" `

aduth commented on packages/components/src/slot-fill/bubbles-virtually/slot-fill-provider.js:
```diff
@@ -0,0 +1,92 @@
+			const { [ name ]: slot, ...nextSlots } = prevSlots;
+			// Make sure we're not unregistering a slot registered by another element
+			// See https://github.com/WordPress/gutenberg/pull/19242#issuecomment-590295412
+			if ( slot.ref === ref ) {
```
Should it be safe to assume `slot` is defined here?

In debugging the issue mentioned by @draganescu at https://github.com/WordPress/gutenberg/issues/18560#issuecomment-604920204 , the exact error occurs at this line of code:

![image](https://user-images.githubusercontent.com/1779930/77765354-874f4f00-7014-11ea-97a2-54cf7c32cc24.png)

It's not yet clear to me if the bug is in how the slot came to be `undefined` in the first place, or if we should be guarded in how we expect `slot` to be available at this line of code.
